extension-qr-code/src/org/eclipse/birt/extension/qrcode/QRCodeBuilder.java

```diff
@@ -0,0 +1,191 @@
+
+		newShell.setText("QRCode Builder");
+	}
+
+	private void openExpression(Text textControl) {
+		String oldValue = textControl.getText();
+
+		ExpressionBuilder eb = new ExpressionBuilder(textControl.getShell(), oldValue);
+		eb.setExpressionProvider(new ExpressionProvider(qrItem.getModelHandle()));
+
+		String result = oldValue;
+
+		if (eb.open() == Window.OK) {
+			result = eb.getResult();
+		}
+
+		if (!oldValue.equals(result)) {
+			textControl.setText(result);
+		}
+	}
+
+	@Override
+	protected Control createDialogArea(Composite parent) {
+		Composite composite = new Composite(parent, SWT.NONE);
+		GridLayout layout = new GridLayout(3, false);
+		layout.marginHeight = convertVerticalDLUsToPixels(IDialogConstants.VERTICAL_MARGIN);
+		layout.marginWidth = convertHorizontalDLUsToPixels(IDialogConstants.HORIZONTAL_MARGIN);
+		layout.verticalSpacing = convertVerticalDLUsToPixels(IDialogConstants.VERTICAL_SPACING);
+		layout.horizontalSpacing = convertHorizontalDLUsToPixels(IDialogConstants.HORIZONTAL_SPACING);
+		composite.setLayout(layout);
+		composite.setLayoutData(new GridData(GridData.FILL_BOTH));
+
+		lbText = new Label(composite, SWT.None);
+		lbText.setText("Text Content:"); //$NON-NLS-1$
+
+		txtText = new Text(composite, SWT.BORDER);
+		GridData gd = new GridData(GridData.FILL_HORIZONTAL);
+		txtText.setLayoutData(gd);
+
+		Button btnExp = new Button(composite, SWT.PUSH);
+		btnExp.setText("..."); //$NON-NLS-1$
+		btnExp.setToolTipText("Invoke Expression Builder");
+
+		btnExp.addSelectionListener(new SelectionAdapter() {
+
+			@Override
+			public void widgetSelected(SelectionEvent event) {
+				openExpression(txtText);
+			}
+		});
+
+		lbDotsWidth = new Label(composite, SWT.None);
+		lbDotsWidth.setText("Width (dots):");
+
+		spDotsWidth = new Spinner(composite, SWT.None);
+		spDotsWidth.setMinimum(21);
+		spDotsWidth.setMaximum(2000);
+		spDotsWidth.setDigits(0);
+		gd = new GridData(GridData.FILL_HORIZONTAL);
+		gd.horizontalSpan = 2;
+		spDotsWidth.setLayoutData(gd);
+
+		lbEncoding = new Label(composite, SWT.None);
+		lbEncoding.setText("Encoding:");
+
+		txtEncoding = new Text(composite, SWT.None);
+		txtEncoding.setLayoutData(gd);
+
+		applyDialogFont(composite);
+
+		initValues();
+
+		return composite;
+	}
+
+	private void initValues() {
+		txtText.setText(qrItem.getText());
+		spDotsWidth.setSelection(qrItem.getDotsWidth());
+		txtEncoding.setText(qrItem.getEncoding());
+	}
+
+	protected static final Logger logger = Logger.getLogger(QRCodeBuilder.class.getName());
+
+	@Override
+	protected void okPressed() {
+
+			try {
+				qrItem.setText(txtText.getText());
+				qrItem.setEncoding(txtEncoding.getText());
+				qrItem.setDotsWidth(Integer.parseInt(spDotsWidth.getText()));
+				super.okPressed();
+			} catch (SemanticException e) {
+				// Should not happen, because the dialog controls should prevent invalid data.
+				// If it happens nevertheless, we log this and we do not leave the dialog.
```

Comment:
Sorry, but this is not the way how JFace Dialog should work. No `try-catch` should be in `okPressed`, it is too late to validate. We have button `enabled` state to control the ability to "apply and close the dialog", controls should be validated on value change and the result should be propagated to the `enabled` state